### PR TITLE
Replaces smileys and the logic of replacing links/emails.

### PR DESCRIPTION
**Pull request by damencho**
Now using react-emoji-render and react-linkify.

### CODE REVIEW

**Comment by damencho** [comment]
@saghul Sadly this is not available in the latest released version :(

**Comment by damencho** [comment]
But there is a way to add those :) A fix is comming

File: css/_chat.scss [hunk 1]
```diff
@@ -214,12 +214,11 @@
...
 
 .smiley {
-    height: 26px;
+    font-size: 14pt;
 }
 
 #smileys {
-    background: white;
-    border-radius: 50%;
+    font-size: 20pt;
```
Comment (virtuacoplenny) [inline css/_chat.scss]: Why did you go with pts over px (or em)?

Comment (damencho) [reply css/_chat.scss]: Hum, no reason, probably I was looking at some other css values while trying to figure out the correct size. Should I change it?

**Comment by damencho** [comment]
> I noticed a gap between the smiley picker and the input. Intentional?
Nope. I did as fewer changes as possible ...

File: react/features/chat/components/ChatMessage.web.js [hunk 2]
```diff
@@ -61,7 +62,22 @@ class ChatMessage extends PureComponent<Props> {
         const escMessage = messagetoDisplay.replace(/</g, '&lt;')
             .replace(/>/g, '&gt;')
             .replace(/\n/g, '<br/>');
-        const messageWithHTML = processReplacements(escMessage);
+        const processedMessage = [];
+        const items = toArray(escMessage, { className: 'smiley' });
+
+        items.forEach(i => {
```
Comment (virtuacoplenny) [inline react/features/chat/components/ChatMessage.web.js]: You can use a map to express iterating over items and returning something for each. 

Do you think it's obvious enough that toArray returns the string split at potential smileys? I'm a little concerned that "items" is a vague name for the return of toArray. Maybe this is more a criticism of toArray, though.

File: css/_chat.scss (hunk 1 above)
Comment (virtuacoplenny) [reply css/_chat.scss]: If you had not reason to use it then I do suggest you change it to use px or em. em scales relative to the parent, which is maybe what we want in the future (like 15 years from now, let's be honest). The easier would be px and that's the most common measurement used in the jitsi css.

**Comment by virtuacoplenny** [comment]
> > Nope. I did as fewer changes as possible ...
Looks like the CSS animation works based on height. When hidden, the panel has a height of 0. Then there's a hardcode height that the panels is set to (202px) when the panel is to be shown. The 202px is too tall now. If you don't want to alter the animation, the easiest fix would be change the height declaration for .smileys-panel.show-smileys.

**Comment by virtuacoplenny** [comment]
Just noticed the smileys are no longer center aligned. There are a couple quick fixes. For example, remove the 7px padding from .smiley, set .smileyContainer to have text-align: center, and pixel push a little as needed.
alignment

File: css/_chat.scss (hunk 1 above)
Comment (damencho) [reply css/_chat.scss]: One thing I notice is that the whole _chat.scss has `pt` in it, it is not just this change.

File: react/features/chat/components/ChatMessage.web.js (hunk 2 above)
Comment (damencho) [reply react/features/chat/components/ChatMessage.web.js]: This is basically an example code about Linkify I was given and I used. I will try now to replace it with something more readable.

**Comment by damencho** [comment]
Hey @virtuacoplenny can you make another round, when you have a chance. 
- For the toArray thing, I just added a comment (copied it from the example of react-emoji-render) and changed the variable name.
- For the emoji popup I just edited the height to match the current one...
- For centering every emoji I took your recommendations and also edited the container height to match the height of the inner components.
- I haven't edited the `font-size: 20pt;` as the whole scss is full of pt values, and if you say to change it, we need to change all of them, wdyt?

**Comment by virtuacoplenny** [comment]
We need to change them all for the sake of consistency yeah, but if it's a wide enough problem then I agree on holding off and not doing it for this commit.

**Comment by saghul** [comment]
🎉 🎉 🎉  thanks for walking the extra mile @damencho

**Pull request merged**